### ISSUE TITLE
Can't have same port name

### ISSUE THREAD

**Issue by tianxu0836** [clientIO/joint#347]
If the inport and outports contains a same name, it will not work!
i.e.
This will not work:
  inPorts: ['aaa', '.bbb'],
  outPorts: ['aaa', 'bbb']

Here is my code

[http://jsfiddle.net/tianxu0836/L2f73cbf/50/](http://jsfiddle.net/tianxu0836/L2f73cbf/50/)

`var graph = new joint.dia.Graph;
var paper = new joint.dia.Paper({
  el: $('#paper'),
  width: 600,
  height: 600,
  gridSize: 1,
  model: graph,
  defaultLink: new joint.dia.Link({
    attrs: {
      '.connection': {
        stroke: '#1E4B79'
      },
      '.marker-target': {
        fill: '#1E4B79',
        stroke: '#1E4B79',
        d: 'M 10 0 L 0 5 L 10 10 z'
      }
    }
  }),
  validateConnection: function(cellViewS, magnetS, cellViewT, magnetT, end, linkView) {
    // Prevent linking from input ports.
    if (magnetS && magnetS.getAttribute('type') === 'input') return false;
    // Prevent linking from output ports to input ports within one element.
    if (cellViewS === cellViewT) return false;
    // Prevent linking to input ports.
    return magnetT && magnetT.getAttribute('type') === 'input';
  },
  // Enable marking available cells & magnets
  markAvailable: true
});

joint.shapes.devs.TopBottomModel = joint.shapes.basic.Generic.extend(_.extend({}, joint.shapes.basic.PortsModelInterface, {
  markup: '<g class="rotatable"><g class="scalable"><rect class="body"/></g><text class="label"/><g class="inPorts"/><g class="outPorts"/></g>',
  portMarkup: '<g class="port port<%= id %>"><rect class="port-body"/><text class="port-label"/></g>',
  defaults: joint.util.deepSupplement({
    type: 'devs.TopBottomModel',
    size: {
      width: 1,
      height: 1
    },
    inPorts: [],
    outPorts: [],
    attrs: {
      '.': {
        magnet: false
      },
      '.body': {
        width: 350,
        height: 250,
        stroke: 'white',
        fill: 'white',
        rx: '15',
        ry: '15'
      },
      '.port-body': {
        width: 70,
        height: 30,
        magnet: true,
        stroke: '#D8D8D8',
        fill: '#D8D8D8',
        rx: '5',
        ry: '5'
      },
      '.inPorts .port-body': {
        type: 'input'
      },
      '.outPorts .port-body': {
        type: 'output'
      },
      text: {
        fill: 'black',
        'pointer-events': 'none'
      },
      '.label': {
        text: 'App\nName',
        'ref-x': 10,
        'ref-y': .2,
        'ref': '.body'
      },
      // CHANGED: find better positions for port labels 
      '.inPorts .port-label': {
        'ref-x': .14,
        'ref-y': .155,
        ref: 'rect',
        'y-alignment': 'middle',
        'x-alignment': 'middle'
      },
      '.outPorts .port-label': {
        'ref-x': .14,
        'ref-y': .155,
        ref: 'rect',
        'y-alignment': 'middle',
        'x-alignment': 'middle'
      }
      //
    }

  }, joint.shapes.basic.Generic.prototype.defaults),

  getPortAttrs: function(portName, index, total, selector, type) {

```
var attrs = {};

var portClass = 'port' + index;
var portSelector = selector + '>.' + portClass;
var portLabelSelector = portSelector + '>.port-label';
var portBodySelector = portSelector + '>.port-body';

attrs[portLabelSelector] = {
  text: portName
};
attrs[portBodySelector] = {
  port: {
    id: portName || _.uniqueId(type),
    type: type
  }
};

// CHANGED: swap x and y ports coordinates ('ref-y' => 'ref-x')
attrs[portSelector] = {
  ref: '.body',
  'ref-x': (index + 0.2) * (1 / total)
};
// ('ref-dx' => 'ref-dy')
if (selector === '.outPorts') {
  attrs[portSelector]['ref-dy'] = 0;
}
if (selector === '.inPorts') {
  attrs[portSelector]['ref-y'] = -30;
}
//

return attrs;
```

  }
}));

joint.shapes.devs.TopBottomModelView = joint.shapes.devs.ModelView;

var model = new joint.shapes.devs.TopBottomModel({
  position: {
    x: 50,
    y: 50
  },
  size: {
    width: 260,
    height: 100
  },
  inPorts: ['.tg.gz', '.bams'],
  outPorts: ['.bam', '.bai']
});

var model2 = new joint.shapes.devs.TopBottomModel({
  position: {
    x: 300,
    y: 400
  },
  size: {
    width: 260,
    height: 100
  },
  inPorts: ['5 5 5', '.bams'],
  outPorts: ['.bam', '.bai']
});

graph.addCell(model);
graph.addCell(model2);
`

**Comment by vtalas** [comment]
We've been finishing a new API for manipulation with ports, restriction you've described - among others - is sorted out there. Approach of creating ports by exentendig `joint.shapes.basic.PortsModelInterface,` is going to be marked as deprecated.

**Comment by tianxu0836** [comment]
Even the example in "Working with Ports" is not working with this case.

What is the solution?

It won't make sense that input can't have same name with output....


`(function() {

    var graph = new joint.dia.Graph;
    var paper = new joint.dia.Paper({
        el: $('#paper'),
        width: 650, height: 200, gridSize: 1,
        model: graph,
        defaultLink: new joint.dia.Link({
            attrs: { '.marker-target': { d: 'M 10 0 L 0 5 L 10 10 z' } }
        }),
        validateConnection: function(cellViewS, magnetS, cellViewT, magnetT, end, linkView) {
            // Prevent linking from input ports.
            if (magnetS && magnetS.getAttribute('type') === 'input') return false;
            // Prevent linking from output ports to input ports within one element.
            if (cellViewS === cellViewT) return false;
            // Prevent linking to input ports.
            return magnetT && magnetT.getAttribute('type') === 'input';
        },
        // Enable marking available cells & magnets
        markAvailable: true
    });

    var m1 = new joint.shapes.devs.Model({
        position: { x: 50, y: 50 },
        size: { width: 90, height: 90 },
        inPorts: ['a','b'],
        outPorts: ['a','b'],
        attrs: {
            '.label': { text: 'Model', 'ref-x': .4, 'ref-y': .2 },
            rect: { fill: '#2ECC71' },
            '.inPorts circle': { fill: '#16A085', magnet: 'passive', type: 'input' },
            '.outPorts circle': { fill: '#E74C3C', type: 'output' }
        }
    }).addTo(graph);

    var m2 = m1.clone().translate(300, 0).attr('.label/text', 'Model 2').addTo(graph);

})()`

**Comment by tianxu0836** [comment]
Can anyone provide a working example for me to support this case?

**Comment by kumilingus** [comment]
It has been resolved in `v1.0.1`. There is a separate logic for port `ids` and `labels` now. Please refer to JointJS [docs](http://resources.jointjs.com/docs/jointjs/v1.0/joint.html#dia.Element.ports).

Note this does not apply for `joint.shapes.devs.Model` as this aims to simplify the ports API only and keep the backwards compatibility. Overriding this [method](https://github.com/clientIO/joint/blob/master/plugins/shapes/joint.shapes.devs.js#L109-L120) would be also a way to go though.
